cloudinit/sources/DataSourceTencentCloud.py

```diff
@@ -0,0 +1,259 @@
+
+    def add_config_mod_always(self, mod_name):
+        modules = self.sys_cfg.get("cloud_config_modules", [])
+        modules.append([mod_name, "always"])
+        self.cfg["cloud_config_modules"] = modules
+
+    def check_item_change(self, item, value):
+        if not value and item != "ssh_key":
+            return False
+        changed = False
+        d_item = {}
+        item_path = "/var/lib/cloud/tencentcloud.item"
+        try:
+            if os.path.exists(item_path):
+                with open(item_path, "r") as f:
+                    d_item = json.load(f)
+                if d_item.get(item) != value:
+                    changed = True
+            else:
+                changed = True
+            if changed:
+                d_item[item] = value
+                with open(item_path, "w") as f:
+                    json.dump(d_item, f)
+        except Exception as e:
+            LOG.debug(str(e))
+            LOG.debug(traceback.format_exc())
+            changed = True
+        LOG.debug("%s changed %s", item, changed)
+        return changed
+
+    def del_authorized_keys(self, old_entries, keys):
+        to_del = []
+        for i in range(0, len(old_entries)):
+            ent = old_entries[i]
+            if not ent.valid():
+                continue
+            # Replace those with the same base64
+            for k in keys:
+                if k.base64 == ent.base64:
+                    # Replace it with our better one
+                    ent = k
+                    to_del.append(k)
+            old_entries[i] = ent
+
+        # Now format them back to strings...
+        lines = [str(b) for b in old_entries if b not in to_del]
+
+        # Ensure it ends with a newline
+        lines.append("")
+        return ("\n".join(lines), len(to_del) > 0)
+
+    def del_user_keys(self, keys, username, options=None):
+        # Make sure the users .ssh dir is setup accordingly
+        (ssh_dir, pwent) = ssh_util.users_ssh_info(username)
+        if not os.path.isdir(ssh_dir):
+            util.ensure_dir(ssh_dir, mode=0o700)
+            util.chownbyid(ssh_dir, pwent.pw_uid, pwent.pw_gid)
+
+        # Turn the 'update' keys given into actual entries
+        parser = ssh_util.AuthKeyLineParser()
+        key_entries = []
+        for k in keys:
+            key_entries.append(parser.parse(str(k), options=options))
+
+        # Extract the old and make the new
+        (auth_key_fn, auth_key_entries) = ssh_util.extract_authorized_keys(
+            username
+        )
+        with util.SeLinuxGuard(ssh_dir, recursive=True):
+            content, to_del = self.del_authorized_keys(
+                auth_key_entries, key_entries
+            )
+            if to_del:
+                util.ensure_dir(os.path.dirname(auth_key_fn), mode=0o700)
+                util.write_file(auth_key_fn, content, mode=0o600)
+                util.chownbyid(auth_key_fn, pwent.pw_uid, pwent.pw_gid)
+
+    def get_data(self):
```

Comment:
I'm really confused about what's happening in this function

Reply:
this function gets metadata of the Tencent Cloud instance,like hostname, SSH keys, passwords, etc.and stores them in "self.cfg" for use by other moules and functions

Reply:
Most of the things being done in this method shouldn't really be done by a datasource. You're potentially replacing data provided by the user with meta data from the IMDS. Cloud-init's generally philosophy is that the user should always have the last say, and this prevents them from doing so. For example, if the user sets their own runcmd, how can that work if you're overwriting it with runcmd data from the IMDS? For this, I would suggest checking if the key exists in userdata, and if so, use it. Otherwise use what you have in your metadata.

I have the same concern with modifying the module definitions in `/etc/cloud/cloud.cfg`. Having `runcmd` get run every boot is going to be confusing from somebody who has used cloud-init on any other cloud, and they don't really have the option to change it since you're changing it here on every boot.

There are generally two ways for clouds to set defaults:
- Modify your OS images to create config overrides in `/etc/cloud/cloud.cfg.d/`. Much of the changes shown here could instead be put there.
- Provide [vendor data](https://cloudinit.readthedocs.io/en/latest/explanation/vendordata.html) rather than overwriting any provided user data.